daemon/src/maker_cfd_actor.rs

```diff
@@ -1,216 +1,344 @@
-                                        )
-                                        .unwrap()
-                                }
-                            },
-                            setup_contract_actor::OwnParams::Maker(maker_params),
-                            sk,
-                            oracle_pk,
-                            cfd,
-                        );
-
-                        tokio::spawn({
-                            let sender = sender.clone();
-
-                            async move {
-                                sender
-                                    .send(Command::CfdSetupCompleted(actor.await))
-                                    .unwrap()
-                            }
-                        });
-                        current_contract_setup = Some(inbox);
-                    }
-                    maker_cfd_actor::Command::IncProtocolMsg(msg) => {
-                        let inbox = match &current_contract_setup {
-                            None => panic!("whoops"),
-                            Some(inbox) => inbox,
-                        };
-
-                        inbox.send(msg).unwrap();
-                    }
-                    maker_cfd_actor::Command::CfdSetupCompleted(_finalized_cfd) => {
-                        todo!("but what?")
-                    }
+    takers: Option<Address<MakerIncConnectionsActor>>,
+    current_order_id: Option<OrderId>,
+    current_contract_setup: Option<mpsc::UnboundedSender<SetupMsg>>,
+}
+impl xtra::Actor for MakerCfdActor {}
+
+impl MakerCfdActor {
+    pub async fn new(
+        db: sqlx::SqlitePool,
+        wallet: Wallet,
+        oracle_pk: schnorrsig::PublicKey,
+        cfd_feed_actor_inbox: watch::Sender<Vec<Cfd>>,
+        order_feed_sender: watch::Sender<Option<Order>>,
+        wallet_feed_sender: watch::Sender<WalletInfo>,
+    ) -> Self {
+        let mut conn = db.acquire().await.unwrap();
+        // populate the CFD feed with existing CFDs
+        cfd_feed_actor_inbox
+            .send(load_all_cfds(&mut conn).await.unwrap())
+            .unwrap();
+
+        Self {
+            db,
+            wallet,
+            oracle_pk,
+            cfd_feed_actor_inbox,
+            order_feed_sender,
+            wallet_feed_sender,
+            takers: None,
+            current_order_id: None,
+            current_contract_setup: None,
+        }
+    }
+
+    async fn handle_new_order(&mut self, msg: NewOrder) -> anyhow::Result<()> {
+        let order = msg.0;
+
+        // 1. Save to DB
+        let mut conn = self.db.acquire().await.unwrap();
+        insert_order(&order, &mut conn).await.unwrap();
+
+        // 2. Update actor state to current order
+        self.current_order_id.replace(order.id);
+
+        // 3. Notify UI via feed
+        self.order_feed_sender.send(Some(order.clone()))?;
+
+        // 4. Inform connected takers
+        self.takers
+            .as_ref()
+            .expect("actor to be initialized")
```

Comment:
You can use `.context` to go from `Option` -> `anyhow::Result` which you can then `?`.